# Long-read Single-Cell Analysis Pipeline (ONT + PIPseq)

This pipeline was used to analyze single-cell data generated by an adapted PIPseq protocol for Oxford Nanopore Technologies (ONT) long-read sequencing. Please note that this Github utilizes count matrices that have first undergone our Single-Cell Nextflow Pipeline, which can be found in the following GitHub repository: (https://github.com/UK-SBCoA-EbbertLab/single_cell_nextflow_pipeline).

**Read our paper for more information about these pipelines: https://www.biorxiv.org/content/10.1101/2025.10.16.682832v1**

## Analysis Overview
This analysis pipeline consists of 9 steps, including one which consists of tuning parameters for analyses. These analyses were performed in a singularity container, as detailed in our paper, which can be retrieved as seen in the singularity_container directory.

### Step 0 -- Initial Filtering
This step reduces file-size for more efficient downstream performance by removing genes and isoforms expressed in very few cells, and cells likely representing empty droplets, potentially with ambient RNA. The resulting filtered matrices are then rounded to two decimal places to further minimize file size and transposed for compatibility with downstream steps. 

This step is implimented in Java. For setup and usage details, refer to dedicated README.md in the Step 0 directory.

### Step 1 -- Count Matrix Preparation
This step annotates the expression matrices produced in Step 0 and prepares the structure for downsteam analyses. The GRCh38.113 GTF is parsed to add in the gene symbol according to ensembl ID, and a combined ID with gene symbol, ensembl gene ID, and ensembl transcript ID (Gene-level-- SYMBOL:ENSGID; Isoform-level--SYMBOL:ENSGID:ENSTID. In the case of novel genes, the absent symbol is omitted (BambuGene# or BambuGene#:BambuTx#). These annotated matrices are written to Parquet matrices and stored as an AnnData object (.h5ad) for efficient storage and loading.

### Step 2 -- Quality Control Filtering
This step loads the .h5ad matrices from Step 1 and computes per-cell, per-gene, and per-isoform quality control metrics. Cells are filtered based on the thresholds determined for total reads, % mitochondrial genes/isoforms, number of unique genes/isoforms, and % hemoglobin genes/isoforms per cell (see Table S2). Genes and isoforms with low expression in both replicates (<10 cells) were removed. Doublet detection and removal are performed using the Scrublet package and datasets were concatenated for downstream modeling. Optional summary statistics are also generated as shown in Table S7.

### Step 3 -- AutoZI Parameter Tuning (optional)
These scripts were used to identify which parameters best modeled our PBMC dataset based on convergence of training and validation ELBO curves, batch mixing and T cell marker expression cluster distinction. These are split by gene and isoform parameter tuning for efficiency, since AutoZI can be computationally intensive.

### Step 4 -- AutoZI Modeling
Based on the parameters determined from Step 3, the concatenated datasets from Step 2 are submitted as a job for AutoZI modeling via these python scripts. 

### Step 5 -- Cell-Type Assignment
This step individually plots potential cell-type markers, from known flow cytometry markers of each cell-type. This script shows the process we went through to narrow down appropriate markers from our starting list. We then used the determined markers to assign cell-type and created Figures 3a-b and 3d-m from our manuscript. 

### Step 6 -- Sub-Cell Type Assignment
This step separated out the T cells from the main dataset and determined sub-cell-types based on known flow cytometry markers of each cell-type. This script shows the process we underwent to narrow down appropriate markers from our starting list. We then used the determined markers to assign cell-type and created Figures 5a-b and 5f-m. 

### Step 7 -- Bulk Figure Generation

### Step 8 -- Paper Figure Generation

### Step 9 -- Unique / Raw Count Analyses
